dev/src/bulk-writer.ts

```diff
@@ -582,15 +600,63 @@ export class BulkWriter {
       {lastUpdateTime?: Timestamp} | unknown | string | FieldPath
     >
   ): Promise<WriteResult> {
-    this.verifyNotClosed();
-    const bulkCommitBatch = this.getEligibleBatch();
-    const resultPromise = bulkCommitBatch.update(
-      documentRef,
-      dataOrField,
-      ...preconditionOrValues
+    return this.runOperation(documentRef, 'set', bulkCommitBatch =>
+      bulkCommitBatch.update(documentRef, dataOrField, ...preconditionOrValues)
     );
-    this.sendReadyBatches();
-    return resultPromise;
+  }
+
+  /**
+   * Attaches a listener that is run every time a BulkWriter operation
+   * successfully completes.
+   *
+   * @param callback A callback to be called every time a BulkWriter operation
+   * successfully completes.
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   *
+   * bulkWriter
+   *   .onWriteResult((documentRef, result) => {
+   *     console.log(
+   *       'Successfully executed write on document: ',
+   *       documentRef,
+   *       ' at: ',
+   *       result
+   *     );
+   *   });
+   */
+  onWriteResult(
+    callback: (
+      documentRef: firestore.DocumentReference,
+      result: WriteResult
+    ) => void
+  ): void {
+    this._successFn = callback;
+  }
+
+  /**
+   * Attaches a listener that is run every time a BulkWriter operation fails.
+   *
+   * @param shouldRetryCallback A callback to be called every time a BulkWriter
+   * operation fails. Returning `true` will retry the operation. Returning
+   * `false` will stop the retry loop.
+   * @example
+   * let bulkWriter = firestore.bulkWriter();
+   *
+   * bulkWriter
+   *   .onWriteError((error) => {
+   *     if (
+   *       shouldRetry(error.code) &&
```

Comment:
Can you just use some hardcoded error check (`error.code === GrpcStatus.UNAVAILABLE`) in this example?